### ISSUE TITLE
Zinnia0.18, in Python3 venv, DEBUG on, Cannot resolve keyword 'entries' into field. Choices are:avatar, comment_comments,...

### ISSUE THREAD

**Issue by wcxaaa** [Fantomas42/django-blog-zinnia#484]
Hi!
I'm using **Django 1.10** and I'm trying to deploy blog Zinnia in a **Python 3 virtualenv**. 
I installed **Zinnia 0.18** using pip, and followed the "get started" guide to set everything in my site settings.py and urls.py, then made the migrations.
However, when I went to 127.0.0.1/weblog , the exception raised.

> django.core.exceptions.FieldError: Cannot resolve keyword 'entries' into field. Choices are: avatar, comment_comments, comment_flags, date_joined, email, first_name, groups, id, is_active, is_staff, is_superuser, last_login, last_name, logentry, password, user_permissions, username

I looked up issues solved (or closed) before, but unluckily none of them works for me.

Apps I installed:

Django==1.10
Pillow==3.3.1
PyMySQL==0.7.6
beautifulsoup4==4.5.1
django-appconf==1.0.2
django-avatar==3.0.0
django-blog-zinnia==0.18
django-contrib-comments==1.7.2
django-mptt==0.8.6
django-tagging==0.4.5
django-xmlrpc==0.1.5
mots-vides==2015.5.11
pyparsing==2.1.9
pytz==2016.6.1
regex==2016.08.27

**Comment by Fantomas42** [comment]
Hello,

do you use a custom model User ?

**Comment by wcxaaa** [comment]
No. Users are now managed by Django admin itself. 
I used my model only in home app, which includes 2 classes, with nothing about users.

**Comment by wcxaaa** [comment]
I start a pure new project, with only Zinnia enabled, using Default SQLite Database.

This issue remains. :(

> Cannot resolve keyword 'entries' into field. Choices are: comment_comments, comment_flags, date_joined, email, first_name, groups, id, is_active, is_staff, is_superuser, last_login, last_name, logentry, password, user_permissions, username

**Comment by wcxaaa** [comment]
I then create a new Python3 virtualenv, install only Django 1.10 and Zinnia.

I start a new project, and set up everything in settings.py and urls.py, use default SQLite Database.

However, this issue still remains.

> django.core.exceptions.FieldError: Cannot resolve keyword 'entries' into field. Choices are: comment_comments, comment_flags, date_joined, email, first_name, groups, id, is_active, is_staff, is_superuser, last_login, last_name, logentry, password, user_permissions, username

Could that be a problem of Python 3 venv, or Django version (because not the latest)?

Or still I missed something?

Should I post the trace logs?

**Comment by Fantomas42** [comment]
This issue seems related to the associated between User and Entry model.
But I don't know why...

Can you paste me your urls and settings files ?

**Comment by wcxaaa** [comment]
There are too many lines in these two files, so I pack them up.
It's a pure new test Django project, with this issue inside.

https://app.box.com/s/034jq6by8fkqshmz0k9bmzutcnxudun7

Thanks.

**Comment by Fantomas42** [comment]
Okay for the now, the bug seems coming from Python3 only.
I'm investigating.

**Comment by Fantomas42** [comment]
Okay, I don't know why but using DEBUG = False seems solving the issue on Python 3

**Comment by wcxaaa** [comment]
Setting DEBUG = False does temporarily solve the keyword resolution exception. 
When I visit 127.0.0.1:8000/weblog again, it raises no exception.

But everything on the page is plain, with no style-sheet applied. It also says "**Error 400 bad request**".

**Comment by wcxaaa** [comment]
More strangely, when I set DEBUG = False, my own index page disappears.
127.0.0.1:8000/ leads directly to Zinnia blog ( with **Error 400 bad request** ), rather than my own index page.
I'm a bit new to Django, so I don't know why about this.

**Comment by Fantomas42** [comment]
> But everything on the page is plain, with no style-sheet applied.

This behaviour is normal, when DEBUG is False, the static files are not handled by Django.

**Comment by bittner** [comment]
In other words, look at the [Django deployment checklist](https://docs.djangoproject.com/en/stable/howto/deployment/checklist/). Specifically,

- [DEBUG](https://docs.djangoproject.com/en/stable/howto/deployment/checklist/#debug)
- [ALLOWED_HOSTS](https://docs.djangoproject.com/en/stable/howto/deployment/checklist/#allowed-hosts)
- [Deployment with staticfiles](https://docs.djangoproject.com/en/stable/howto/static-files/#deployment) (the `collectstatic` management command)

**Comment by wcxaaa** [comment]
Thanks, I'll check the static files part later, but what about the "**Error 400 bad request**", and the **misleading** issue?

**Comment by Fantomas42** [comment]
The Error 400 is certainly related to the ALLOWED_HOSTS.

About the misleading issue, I'm working on it, so be patient.

And note, I don't mind fixing bugs, but here is not your personal support center,
if you want to solve the issue quicker, your help would be appreciated.

**Comment by wcxaaa** [comment]
Thanks. But literally I'm not intentionally using it as a client center, and sorry I'm not actually pushing you guys as it may seem, but no.
 I'm a bit new to Django, so my help would be limited, but I'm doing the best I can.

About the misleading issue, I guess it may be the reason of the previous **Error 400 bad request**, as if I turn the DEBUG on again, my own home page will show again. Besides, on the mislead-ed page, it also has the word **Error 400 bad request** shown in the center.

**Comment by bittner** [comment]
Please, do us a favor:

- Click on the "ALLOWED_HOSTS" link above, read the explanations and try to understand them. This is critical for advancing in understanding Django (for taking it to production).
- There is a link over there to the [ALLOWED_HOSTS settings]/https://docs.djangoproject.com/en/stable/ref/settings/#std:setting-ALLOWED_HOSTS) description. Read also that.

Your solution will look something like this, probably: (for local development)
``` python
ALLOWED_HOSTS = [
    '127.0.0.1',
    'localhost',
]
```
But don't trust me! Understand why you need to do this, then take action.

**Comment by wcxaaa** [comment]
Okay, Zinnia page successfully opened just now. Very much appreciated!

(After correctly configured ALLOWED_HOSTS, and runserver by using 
`python3 manage.py runserver --insecure`
)

So the first issue is in DEBUG mod of Python3, I'll change the title now.

**Comment by Fantomas42** [comment]
For the now, I have just understood why the DEBUG setting, enable or disable this bug.

This is related to the admin autodiscovering feature and the checks done on the ModelAdmin classes when DEBUG=True. These checks mess in some way the ORM relation cache system.

At first sight when the applications are ready, the relations are correct, then the admin autodiscover, do some check, and mess with the relations... 

And I still don't know why it's related to Python 3 only.

A good way to reproduce the issue, is to install the project with buildout within a Python 3 virtualenv, then run 

    ./bin/demo test

**Comment by matthieucham** [comment]
Hello Fantomas42,
Thank you very much for investigating this. I have the exact same config as wcxaaa, and run into the exact same issue. I have little experience of Python - Django, however if you lack time I am willing to help.

**Comment by MUDASSARHASHMI** [comment]
I am having the same error with same scenario explained above. But I am using python27.
If there is a solution available please refer to it.
Regards,
Mudassar

**Comment by piepmatz** [comment]
Same here with Python 2.7. `DEBUG=False` works as a workaround.

**Comment by MUDASSARHASHMI** [comment]
Setting DEBUG=False means Django will not present error. So that's not a solution.

**Comment by bittner** [comment]
@MUDASSARHASHMI Are you sure you know what the `DEBUG` setting is about? In this bug this settings seems to have a *side-effect*.

Setting `DEBUG=False` is *not* about hiding errors and bugs. Get yourself an idea from the [documentation](https://docs.djangoproject.com/en/stable/ref/settings/#debug).

**Comment by MUDASSARHASHMI** [comment]
DEBUG=False generally means that it won't provide exceptions generated by the view. Above statement strictly was for the matter in discussion. You can check `[docs ](https://docs.djangoproject.com/en/1.10/ref/settings/#debug)`to confirm.

**Comment by Fantomas42** [comment]
@MUDASSARHASHMI how can I reproduce the error with Python 2.7 ? Because actually I cannot.

I have localized the issue more precisely, but not accurately at this time. 
The good news is that it seems possible to write a patch to fix the situation.

**Comment by MUDASSARHASHMI** [comment]
@Fantomas42 its a wonderful app and it will be unfair of me not to appreciate your effort. You've taken care of everything to detail that might be required for a blog from SEO standpoint as well as extendible for max utilization. But this error I couldn't get rid of. I've gone through all the previous issues related to this topic but couldn't figure out a way within the code set. I've deleted the app to redo it and once I am done with it I will email you to access my virtual server to look into it.

**Comment by Fantomas42** [comment]
The issue seems coming from django / django_comments.

I have found a solution, by registering django_comments after zinnia in the INSTALLED_APPS setting.

Can someone test this fix ?

**Comment by MUDASSARHASHMI** [comment]
It worked on python2 as well as on python3. You rock man. What's the logic?

**Comment by matthieucham** [comment]
I can confirm, it works on Django 1.10 + python 3.4
Thank you Fantomas for this workaround

**Comment by piepmatz** [comment]
Worked here as well, nice.

**Comment by Fantomas42** [comment]
I don't understand what's going under the hood, but here is what I have found.

When `DEBUG` is True, the django.contrib.admin app will apply some check functions on the ModelAdmin class.

When executing the `_check_list_display` on `CommentsAdmin`, this provoke the issue when checking the attribute `'name'` which is a property, not a real field. 

It seems that the check function make a call on` model._meta.get_field(item)`, and destroy the the relationship between the models for a non obvious reason, but certainly related on how the property is calculated (it's a DB call)... 

Making this check call later solve the issue, because the relationship on the Entry model seems OK.

Django_comment should change the CommentsAdmin.list_display to make 'name' a callable, which will defer the check, and do it properly, but I don't think that they will change this now...

So I will the update the documentation.

**Comment by MUDASSARHASHMI** [comment]
So according to the above scenario should it be considered production ready as per your personal evaluation?

**Comment by Fantomas42** [comment]
For me it should be considered production ready, moreover the bug should not appears when `DEBUG` is set to False.

**Comment by Tapo4ek** [comment]
Python 2.7 And Django 1.10.1 - Also doesn't work. DB tried - sqlite and postgres.
```
Cannot resolve keyword 'entries' into field
```

**Comment by MUDASSARHASHMI** [comment]
@Tapo4ek delete db and migrations in all migration folders if you have other apps besides Zinnia. Migrate again and it should work for you.

**Comment by Fantomas42** [comment]
I have made a branch named **feature/fix-author-import**, which should fix the issue in a proper way. 

Can someone test if the issue is still present within this branch or not ? (without changing the order of the apps of course)

**Comment by MUDASSARHASHMI** [comment]
@Fantomas42 Awesome! its working fine now. Thanks for your time and help.

**Comment by Fantomas42** [comment]
@MUDASSARHASHMI thanks for your support. 

I will merge the branch and make a new release.

**Comment by jplehmann** [comment]
I got stung by this when I upgraded to `0.18` while migrating to Django `1.10`. For anyone coming through here:

### TLDR: use `0.18.1` which includes the fix.
